### ISSUE TITLE
Fix issue in mixlib-versioning parsing date based string

### ISSUE THREAD

**Issue by tyler-ball** [chef/chef-workstation#1100]
## Description
Hey y'all. I noticed a 2020.04.14222920+20200414223557 build of Workstation has hit the unstable channel. I decided to try and query that version using the mixlib-install CLI and uncovered a potential error/bug with the new date-based masquerading as SemVer version format. If I directly query for a version of 2020.04.14222920+20200414223557 things work as expected. If we do not provide a version the default of latest is used under the covers a nice error is returned of `undefined method []' for nil:NilClass is returned.
```
➜  mixlib-install download chef-workstation --channel unstable --platform el --platform-version 8 --architecture x86_64 --url --version 0.17.18
https://packages.chef.io/files/unstable/chef-workstation/0.17.18/el/8/chef-workstation-0.17.18-1.el7.x86_64.rpm
➜  mixlib-install download chef-workstation --channel unstable --platform el --platform-version 8 --architecture x86_64 --url --version "2020.04.14222920+20200414223557"
https://packages.chef.io/files/unstable/chef-workstation/2020.04.14222920+20200414223557/el/8/chef-workstation-2020.04.14222920+20200414223557-1.el7.x86_64.rpm
➜  mixlib-install download chef-workstation --channel unstable --platform el --platform-version 8 --architecture x86_64 --url --version latest
undefined method `[]' for nil:NilClass
```
When I fired up IRB and ran the same command as pure Ruby I was able to grab the full Ruby stacktrace behind the `undefined method []' for nil:NilClass error`:
```
irb(main):001:0> require 'mixlib/install'
=> true
irb(main):002:1* mixlib_install_options = {
irb(main):003:1*   channel: :unstable,
irb(main):004:1*   product_name: 'chef-workstation',
irb(main):005:1*   product_version: :latest,
irb(main):006:1*   platform_version_compatibility_mode: true,
irb(main):007:1*   architecture: 'x86_64',
irb(main):008:1*   platform: 'el',
irb(main):009:1*   platform_version: '8'
irb(main):010:0> }
irb(main):011:0> installer = Mixlib::Install.new(mixlib_install_options)
irb(main):012:1* begin
irb(main):013:1*   artifact = installer.artifact_info
irb(main):014:1* rescue Mixlib::Install::Backend::ArtifactsNotFound => e
irb(main):015:1*   abort e.message
irb(main):016:0> end
Traceback (most recent call last):
       16: from /Users/schisamo/.rubies/ruby-2.6.5/lib/ruby/2.6.0/bundler/vendor/thor/lib/thor/invocation.rb:126:in `invoke_command'
       15: from /Users/schisamo/.rubies/ruby-2.6.5/lib/ruby/2.6.0/bundler/vendor/thor/lib/thor/command.rb:27:in `run'
       14: from /Users/schisamo/.rubies/ruby-2.6.5/lib/ruby/2.6.0/bundler/cli.rb:463:in `exec'
       13: from /Users/schisamo/.rubies/ruby-2.6.5/lib/ruby/2.6.0/bundler/cli/exec.rb:28:in `run'
       12: from /Users/schisamo/.rubies/ruby-2.6.5/lib/ruby/2.6.0/bundler/cli/exec.rb:74:in `kernel_load'
       11: from /Users/schisamo/.rubies/ruby-2.6.5/lib/ruby/2.6.0/bundler/cli/exec.rb:74:in `load'
       10: from /Users/schisamo/.gem/ruby/2.6.5/bin/irb:23:in `<top (required)>'
        9: from /Users/schisamo/.gem/ruby/2.6.5/bin/irb:23:in `load'
        8: from /Users/schisamo/.gem/ruby/2.6.5/gems/irb-1.2.3/exe/irb:11:in `<top (required)>'
        7: from (irb):13
        6: from /Users/schisamo/dev/code/chef/mixlib-install/lib/mixlib/install.rb:46:in `artifact_info'
        5: from /Users/schisamo/dev/code/chef/mixlib-install/lib/mixlib/install/backend.rb:29:in `info'
        4: from /Users/schisamo/dev/code/chef/mixlib-install/lib/mixlib/install/backend/base.rb:62:in `info'
        3: from /Users/schisamo/dev/code/chef/mixlib-install/lib/mixlib/install/backend/package_router.rb:43:in `available_artifacts'
        2: from /Users/schisamo/dev/code/chef/mixlib-install/lib/mixlib/install/backend/package_router.rb:113:in `latest_version'
        1: from /Users/schisamo/dev/code/chef/mixlib-install/lib/mixlib/install/backend/package_router.rb:118:in `extract_version_from_response'
NoMethodError (undefined method `[]' for nil:NilClass)
```

When you dig into the affected code you discover the underlying issue appears to be in mixlib-versioning's parsing of a version string like 2020.04.14222920+20200414223557:
```
irb(main):001:0> require 'mixlib/versioning'
=> true
irb(main):002:0> Mixlib::Versioning.parse('0.17.18')
=> #<Mixlib::Versioning::Format::OpscodeSemVer:0x3fd43a43e4fc @major=0, @minor=17, @patch=18, @prerelease=nil, @build=nil, @input="0.17.18">
irb(main):003:0> Mixlib::Versioning.parse('2020.04.14222920')
=> #<Mixlib::Versioning::Format::Rubygems:0x3fd41a023c98 @major=2020, @minor=4, @patch=14222920, @prerelease=nil, @iteration=nil, @input="2020.04.14222920">
irb(main):004:0> Mixlib::Versioning.parse('2020.04.14222920+20200414223557')
=> nil
```

https://github.com/chef/mixlib-versioning

**Comment by clintoncwolfe** [comment]
So, I've traced this to here: https://github.com/chef/mixlib-versioning/blob/876752f1b5aeb2c719a7a193964464ab232ee092/lib/mixlib/versioning/format/semver.rb#L45 which indicates that a leading zero is not permitted in SemVer versions. That's correct per the SemVer spec at https://semver.org/ - see BNF grammar, the section on "numeric identifiers":
```
<numeric identifier> ::= "0"
                       | <positive digit>
                       | <positive digit> <digits>

```

Using `2020.04.14222920+20200414223557` as the example, it's the `04` that's being rejected. If we turn the `+` into a `.`, it correctly parses as-is as a "rubygems" version (since the rubygems parser allows leading zeros in version numbers). Or we can decide that mixlib-versioning can take a loose interpretation of the semver spec.  Thoughts?

**Comment by tyler-ball** [comment]
Trying a build with no zero padding, and using a build number for the 3rd part of the dotted version.

https://buildkite.com/chef/chef-chef-workstation-master-omnibus-adhoc/builds/227#81b3a414-cadd-4bdd-8a4c-8d8afe9d174d

**Comment by tyler-ball** [comment]
We are going to add some tests to `mixlib-versioning` to cover our planned version number of `year.month_no_padding.build_number` where build number is constantly increasing. If those tests do not error we will close out this issue.

**Comment by tyler-ball** [comment]
Closed because we're going with `20.4.x` (short year, non-zero-padded month, build number) which falls within existing symver constraints
